PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx

```diff
@@ -0,0 +1,367 @@
+        xQVec = collision.qvecFT0ARe();
+        yQVec = collision.qvecFT0AIm();
+        break;
+      case qvecEstimator::FT0C:
+        xQVec = collision.qvecFT0CRe();
+        yQVec = collision.qvecFT0CIm();
+      case qvecEstimator::TPCPos:
+        xQVec = collision.qvecBPosRe();
+        yQVec = collision.qvecBPosIm();
+        amplQVec = collision.nTrkBPos();
+        break;
+      case qvecEstimator::TPCNeg:
+        xQVec = collision.qvecBNegRe();
+        yQVec = collision.qvecBNegIm();
+        amplQVec = collision.nTrkBNeg();
+        break;
+      default:
+        LOG(warning) << "Q vector estimator not valid. Please choose between FV0A, FT0M, FT0A, FT0C, TPC Pos, TPC Neg. Fallback to FV0A";
+        xQVec = collision.qvecFV0ARe();
+        yQVec = collision.qvecFV0AIm();
+        break;
+    }
+    return {xQVec, yQVec, amplQVec};
+  }
+
+  /// Compute the scalar product
+  /// \param collision is the collision with the Q vector information and event plane
+  /// \param candidates are the selected candidates
+  template <int DecayChannel, typename T1>
+  void runFlowAnalysis(CollsWithQvecs::iterator const& collision,
+                       T1 const& candidates)
+  {
+    std::vector<float> qVecs = getQvec(collision);
+    float xQVec = qVecs[0];
+    float yQVec = qVecs[1];
+    float amplQVec = qVecs[2];
+    float evtPl = epHelper.GetEventPlane(xQVec, yQVec, harmonic);
+    float cent = getCentrality(collision);
+
+    for (const auto& candidate : candidates) {
+      float massCand = 0.;
+      std::vector<float> outputMl = {-999., -999.};
+
+      if constexpr (std::is_same<T1, CandDsData>::value) {
+        switch (DecayChannel) {
+          case DecayChannel::DsToKKPi:
+            massCand = hfHelper.invMassDsToKKPi(candidate);
+            if (storeMl) {
+              for (unsigned int iclass = 0; iclass < classMl->size(); iclass++)
+                outputMl[iclass] = candidate.mlProbDsToKKPi()[classMl->at(iclass)];
+            }
+            break;
+          case DecayChannel::DsToPiKK:
+            massCand = hfHelper.invMassDsToPiKK(candidate);
+            if (storeMl) {
+              for (unsigned int iclass = 0; iclass < classMl->size(); iclass++)
+                outputMl[iclass] = candidate.mlProbDsToPiKK()[classMl->at(iclass)];
+            }
+            break;
+          default:
+            break;
+        }
+      } else if constexpr (std::is_same<T1, CandDplusData>::value) {
+        massCand = hfHelper.invMassDplusToPiKPi(candidate);
+        if (storeMl) {
+          for (unsigned int iclass = 0; iclass < classMl->size(); iclass++)
+            outputMl[iclass] = candidate.mlProbDplusToPiKPi()[classMl->at(iclass)];
+        }
+      }
+      float ptCand = candidate.pt();
+      float phiCand = candidate.phi();
+
+      // TPC only
+      // If TPC is used for the SP estimation, the tracks of the hadron candidate must be removed from the TPC Q vector
+      // to avoid double counting.
```

Comment:
```suggestion
      // If TPC is used for the SP estimation, the tracks of the hadron candidate must be removed from the TPC Q vector to avoid double counting.
```